In [1]:
# necessery libraries
import pandas as pd
import sys
sys.path.insert(0,'../src/utils/')
import general_look_at_data as  GLAD

# Explore Dataset in General

In [2]:
# data source must be updated to featsel folder rather than raw
data = pd.read_csv('../data/featsel/Citywide_Survey_featsel.csv')

In [3]:
# get the shape of the data
data.shape

(3346, 156)

In [4]:
# set option setting to display all columns
pd.set_option('display.max_columns', None)

In [5]:
# Display the head of data
data.head()

cms_zone  num_trips  num_walk_trips  num_transit_trips  \
0  Inner Brooklyn       23.0             1.0                3.0   
1   Middle Queens       15.0             2.0                1.0   
2   Middle Queens       30.0            22.0                9.0   
3   Middle Queens       48.0            23.0               13.0   
4   Middle Queens       45.0            13.0                6.0   

   num_bike_trips  num_taxi_trips  num_tnc_trips  num_complete_weekend_days  \
0            11.0             0.0            0.0                        2.0   
1             0.0             0.0            0.0                        2.0   
2             0.0             0.0            0.0                        2.0   
3             0.0             0.0            0.0                        2.0   
4            12.0             0.0            2.0                        2.0   

   num_complete_weekdays first_travel_date last_travel_date  made_trips  \
0                    5.0        06/24/2019       06/30/2019         1.0   
1                    5.0        05/29/2019       06/04/2019         1.0   
2                    5.0        06/06/2019       06/12/2019         1.0   
3                    2.0        06/16/2019       06/22/2019         1.0   
4                    5.0        06/08/2019       06/14/2019         1.0   

   smartphone_type  age  employment  worker  student  license  vehicle  \
0                1    5           6       0        1        1        1   
1                2    5           2       1        2        0        1   
2                2    7           7       1        0        0      995   
3                2    6           1       1        0        1        1   
4                2    5           1       1        2        1      995   

   congestion  drive_cbd_freq  planning_apps  job_type  jobs_count  industry  \
0           4               7              1       995         995       995   
1           4               8              3         1           1         8   
2           2             995              3         1           1        15   
3           3               7              1         1           1        16   
4           3             995              1         2           1        13   

   telework_freq   work_cms_zone  missing_work_location  work_mode  \
0            995             NaN                      0        995   
1              8             NaN                      0        100   
2              8             NaN                      1        105   
3              7  Manhattan Core                      0        105   
4              7    Inner Queens                      0        103   

   work_mode_own  work_park_location  work_park_pay  work_vehicle_pass_period  \
0            995                 995            995                       995   
1              6                   6            995                       995   
2            995                 995            995                       995   
3            995                 995            995                       995   
4            995                 995            995                       995   

   work_park_amount_day  work_park_amount_week  work_park_amount_month  \
0                   NaN                    NaN                     NaN   
1                   NaN                    NaN                     NaN   
2                   NaN                    NaN                     NaN   
3                   NaN                    NaN                     NaN   
4                   NaN                    NaN                     NaN   

   work_park_amount_year  work_mode_auto  work_mode_bus  work_mode_rail  \
0                    NaN             995            995             995   
1                    NaN             995            995             995   
2                    NaN             995            995              30   
3                    NaN             995            995              30   
4                    NaN  

In [6]:
# check features data types 
data.dtypes.unique()

array([dtype('O'), dtype('float64'), dtype('int64')], dtype=object)

# Explore Categorical vs. Numerical Datasets

In [7]:
# get numeric and categorical features names 
numeric_features = data.select_dtypes(include=['float64', 'int64'])
categorical_features = data.select_dtypes(include='O')

## Statistical Information from Numerical Dataset

In [8]:
# Display 5 random samples of numeric dataset
numeric_features.sample(5)

num_trips  num_walk_trips  num_transit_trips  num_bike_trips  \
2259       36.0             2.0                0.0             0.0   
2574        4.0             1.0                2.0             0.0   
1239       54.0            45.0               12.0             0.0   
612        59.0             8.0                2.0            15.0   
3004        8.0             8.0                0.0             0.0   

      num_taxi_trips  num_tnc_trips  num_complete_weekend_days  \
2259             0.0            0.0                        2.0   
2574             0.0            1.0                        0.0   
1239             0.0            0.0                        2.0   
612              0.0            3.0                        1.0   
3004             0.0            0.0                        0.0   

      num_complete_weekdays  made_trips  smartphone_type  age  employment  \
2259                    5.0         1.0                2    5           6   
2574                    1.0         1.0                1    5           1   
1239                    5.0         1.0                1    6           1   
612                     4.0         1.0                2    7           1   
3004                    1.0         1.0                0    9           3   

      worker  student  license  vehicle  congestion  drive_cbd_freq  \
2259       0        0        1        1           4               5   
2574       1        0        1        1           4               4   
1239       1        0        1      995           1             995   
612        1        0        1        2           4               7   
3004       1        0        1      995           1             995   

      planning_apps  job_type  jobs_count  industry  telework_freq  \
2259              5       995         995       995            995   
2574              1         1           2        16              6   
1239              1         1           1        11              8   
612               1         1           1        12              2   
3004              8         3           2        15            995   

      missing_work_location  work_mode  work_mode_own  work_park_location  \
2259                      0        995            995                 995   
2574                      0          1            995                 995   
1239                      0        105            995                 995   
612                       0        101            995                 995   
3004                      0        995            995                 995   

      work_park_pay  work_vehicle_pass_period  work_park_amount_day  \
2259            995                       995                   NaN   
2574            995                       995                   NaN   
1239            995                       995                   NaN   
612             995                       995                   NaN   
3004            995                       995                   NaN   

      work_park_amount_week  work_park_amount_month  work_park_amount_year  \
2259                    NaN                     NaN                    NaN   
2574                    NaN                     NaN                    NaN   
1239                    NaN                     NaN                    NaN   
612                     NaN                     NaN                    NaN   
3004                    NaN                     NaN                    NaN   

      work_mode_auto  work_mode_bus  work_mode_rail  work_mode_ferry  \
2259             995            995             995              995   
2574             995            995             995              995   
1239             995            995              30              995   
612               76            995             995              995   
3004             995            995             995              995   

      work_mode_taxi  work_mode_bike  work_bike_park  work_mode_scooter  \
2259      

In [9]:
# Display the columns number of numeric dataset
numeric_features.shape[1]

151

In [10]:
# Display Statistics of numeric data
numeric_features.describe()

num_trips  num_walk_trips  num_transit_trips  num_bike_trips  \
count  3346.000000     3346.000000        3346.000000     3346.000000   
mean     25.540646       10.004782           5.827555        0.419307   
std      21.900434       12.091553           7.646751        2.141812   
min       0.000000        0.000000           0.000000        0.000000   
25%       4.000000        0.000000           0.000000        0.000000   
50%      24.000000        5.000000           2.000000        0.000000   
75%      41.000000       16.000000          10.000000        0.000000   
max     178.000000       93.000000          57.000000       36.000000   

       num_taxi_trips  num_tnc_trips  num_complete_weekend_days  \
count     3346.000000    3346.000000                3346.000000   
mean         0.111775       0.374776                   1.266886   
std          1.664201       1.181070                   0.843984   
min          0.000000       0.000000                   0.000000   
25%          0.000000       0.000000                   0.000000   
50%          0.000000       0.000000                   2.000000   
75%          0.000000       0.000000                   2.000000   
max         92.000000      18.000000                   2.000000   

       num_complete_weekdays   made_trips  smartphone_type          age  \
count            3346.000000  3346.000000      3346.000000  3346.000000   
mean                3.067244     0.907651         1.419008     6.653019   
std                 1.858192     0.289561         0.691730     1.648495   
min                 0.000000     0.000000         0.000000     4.000000   
25%                 1.000000     1.000000         1.000000     5.000000   
50%                 4.000000     1.000000         2.000000     7.000000   
75%                 5.000000     1.000000         2.000000     8.000000   
max                 5.000000     1.000000         3.000000    11.000000   

        employment       worker      student      license      vehicle  \
count  3346.000000  3346.000000  3346.000000  3346.000000  3346.000000   
mean      2.838314     0.705918     0.147041  -411.589659   572.417215   
std       2.267040     0.455697     0.446764  1988.581039   491.458948   
min       1.000000     0.000000     0.000000 -9998.000000     1.000000   
25%       1.000000     0.000000     0.000000     1.000000     1.000000   
50%       1.000000     1.000000     0.000000     1.000000   995.000000   
75%       6.000000     1.000000     0.000000     1.000000   995.000000   
max       7.000000     1.000000     2.000000     1.000000   997.000000   

        congestion  drive_cbd_freq  planning_apps     job_type   jobs_count  \
count  3346.000000     3346.000000    3346.000000  3346.000000  3346.000000   
mean   -409.562463      276.343096    -695.401674   248.800956   248.632098   
std    1989.001956     1610.611102    2551.276292   823.145574   823.196254   
min   -9998.000000    -9998.000000   -9998.000000 -9998.000000 -9998.000000   
25%       2.000000        7.000000       1.000000     1.000000     1.000000   
50%       3.000000        8.000000       4.000000     2.000000     1.000000   
75%       4.000000      995.000000       6.000000   995.000000   995.000000   
max       5.000000      995.000000       8.000000   995.000000   995.000000   

          industry  telework_freq  missing_work_location    work_mode  \
count  3346.000000    3346.000000            3346.000000  3346.000000   
mean    254.481769     367.395696               0.060371   399.687388   
std     821.456099     476.115199               0.238208   426.948825   
min   -9998.000000       1.000000               0.000000     1.000000   
25%       7.000000       7.000000               0.000000   102.000000   
50%      14.000000       8.000000               0.000000   105.000000   
75%     995.000000     995.000000               0.000000   995.000000   
max     995.000000     995.000000               1.000000   995.000000   

       work_mode_own  work_park_lo

### Conclusions Based on the Statistics Presented
* Almost all numerical data is one hot including.
* Missing data values/labels such as -9998 and 995 have an impact on statistics.
* Because work park amount_ and school park amount_ are rare data, their distribution is skewed.

## Statistical Information from Categorical Dataset

In [11]:
# Display 5 samples of categorical dataset
categorical_features.sample(5)

cms_zone first_travel_date last_travel_date   work_cms_zone  \
1244  Northern Manhattan        06/16/2019       06/22/2019             NaN   
619        Staten Island        06/06/2019       06/12/2019   Staten Island   
985   Northern Manhattan        05/31/2019       06/06/2019  Manhattan Core   
3221      Northern Bronx        05/28/2019       05/28/2019             NaN   
1714      Southern Bronx        06/05/2019       06/11/2019             NaN   

     school_cms_zone  
1244             NaN  
619              NaN  
985              NaN  
3221             NaN  
1714             NaN

In [12]:
# Display 5 samples of categorical dataset
categorical_features.shape[1]

5

In [13]:
# Display Statistics of categorical data
categorical_features.describe()

cms_zone first_travel_date last_travel_date   work_cms_zone  \
count             3346              3346             3346            1687   
unique              10                39               40              12   
top     Northern Bronx        06/07/2019       06/13/2019  Manhattan Core   
freq               416               268              279             815   

       school_cms_zone  
count              122  
unique              10  
top     Manhattan Core  
freq                35

### Conclusions based on the statistics presented
* The 'work_cms_zone' feature contain nearly 50% missing data.
* Because 'scool_cms_zone' feature are rare data, its distribution is skewed.

## Summary of Exploring the two Datasets
- There are 151 numeric features but some of them are a kind of labels for the data such as tnc_mode, pakages, etc. 
- There are 5 categorical feature 2 of them are of type time series. 
- The above mentioned wrong data types will be handled during the feature engineering process. 

# Explore Missing Data Information

In [14]:
# Display percentages about types of missing data in the features
feature_missing_stats = GLAD.get_feature_missing(data)
feature_missing_stats

missing_count  missing_percentage  \
feature                                                       
school_park_amount_week            3342            0.998805   
work_park_amount_year              3342            0.998805   
school_park_amount_day             3340            0.998207   
school_park_amount_month           3339            0.997908   
school_park_amount_year            3339            0.997908   
...                                 ...                 ...   
school_bike_park                      0            0.000000   
school_mode_bike                      0            0.000000   
school_mode_taxi                      0            0.000000   
school_egress                         0            0.000000   
disability_none                       0            0.000000   

                          dont_know_percentage  no_response_percentage  \
feature                                                                  
school_park_amount_week                    0.0                     0.0   
work_park_amount_year                      0.0                     0.0   
school_park_amount_day                     0.0                     0.0   
school_park_amount_month                   0.0                     0.0   
school_park_amount_year                    0.0                     0.0   
...                                        ...                     ...   
school_bike_park                           0.0                     0.0   
school_mode_bike                           0.0                     0.0   
school_mode_taxi                           0.0                     0.0   
school_egress                              0.0                     0.0   
disability_none                            0.0                     0.0   

                          technological_error_percentage  \
feature                                                    
school_park_amount_week                              0.0   
work_park_amount_year                                0.0   
school_park_amount_day                               0.0   
school_park_amount_month                             0.0   
school_park_amount_year                              0.0   
...                                                  ...   
school_bike_park                                     0.0   
school_mode_bike                                     0.0   
school_mode_taxi                                     0.0   
school_egress                                        0.0   
disability_none                                      0.0   

                          prefer_not_to_answer_percentage  \
feature                                                     
school_park_amount_week                          0.000000   
work_park_amount_year                            0.000000   
school_park_amount_day                           0.000000   
school_park_amount_month                         0.000000   
school_park_amount_year                          0.000000   
...                                                   ...   
school_bike_park                                 0.000000   
school_mode_bike                                 0.000000   
school_mode_taxi                                 0.000000   
school_egress                                    0.000000   
disability_none                                  0.060072   

                          not_required_percentage feature_type  
feature                                                         
school_park_amount_week                  0.000000      float64  
work_park_amount_year                    0.000000      float64  
school_park_amount_day                   0.000000      float64  
school_park_amount_month                 0.000000      float64  
school_park_amount_year                  0.000000      float64  
...                                           ...          ...  
school_bike_park                         0.997908        int64  
school_mode_bike                         0.997908        int64  
school_mode_taxi  